### PR TITLE
ext_proc: Cache the gRPC client

### PR DESCRIPTION
**Pull request by gbrail**
This is a draft PR for caching the gRPC client -- it will fail the tests until we understand how destructors are being fired by the the gRPC stuff.

Commit Message: Cache the gRPC client to reduce latency in the ext_proc filter
Additional Description:
Risk Level: TBD
Testing: TBD
Docs Changes:
Release Notes:

### CODE REVIEW

**Comment by gbrail** [comment]
@chaoqin-li1123 Can you take a look at this draft PR and give me some advice? I'm trying to figure out why the tests are crashing, and I think it has to do with the order in which the stream is being closed after the filter is destroyed. If the test logs aren't helpful I can run ASAN locally and upload the results. Thanks!

**Comment by chaoqin-li1123** [comment]
Thanks for working on this! I will check out this branch and try to debug it.

**Comment by chaoqin-li1123** [comment]
@htuch It seems that resetStream invoke a defer delete for AsyncStream and it is possible that callback can be invoked after the ExternalProcessorStreamImpl object has been destroyed([this pointer is passed as callback to the grpc client](https://github.com/envoyproxy/envoy/blob/6a97a8c73a94bd4f140be406c81a5ce113695e7a/source/extensions/filters/http/ext_proc/client_impl.cc#L31)). Is there anyway we can work around this?

**Comment by chaoqin-li1123** [comment]
Could you reset the stream_ in onGrpcClose? In the log I found that after grpc finish, we still try to close the stream, which leads to heap_use_after_free error.

**Comment by gbrail** [comment]
I think that we're already resetting the stream. The flow should be, currently:

1. onDestroy is called on the filter
2. it calls ExternalProcessorStreamImpl::close()
3. If we didn't already receive an onRemoteClose() callback on this stream, then we call close() on the stream
4. In this PR, I added code to also call resetStream() here, regardless of whether we called close. (And it seems like we need a close call first or we get into a bad state here.)

However, we still seem to get onRemoteClose() calls after the filter's destructors are called. I'm not sure why that's happening though, since looking at google_async_client_impl.cc we should already be synchronously cleaned up by the time that resetStream completes.

**Comment by chaoqin-li1123** [comment]
Can you free the stream_ in onGrpcClose? This works for me locally.
void Filter::onGrpcClose() {
  ENVOY_LOG(debug, "Received gRPC stream close");
  processing_complete_ = true;
  stats_.streams_closed_.inc();
  // Successful close. We can ignore the stream for the rest of our request
  // and response processing.
  clearAsyncState();
  stream_.reset();
}

In my local test results, asan log reveals that after onGrpcClose, the AsyncStream will be defer deleted, but it is referenced again in the call stack Filter::onDestroy()

**Comment by gbrail** [comment]
Yes, I will try that next. Thanks!

If we free the stream, do you think that we also need to reset it?

**Comment by chaoqin-li1123** [comment]
It seems that the underlying stream will be reset when the onGrpcClose is called. I check all the occurrence of callbacks_.onRemoteClose().

File: source/extensions/filters/http/ext_proc/ext_proc.cc [hunk 1]
```diff
@@ -65,18 +65,26 @@ Filter::StreamOpenState Filter::openStream() {
...
-  // Make doubly-sure we no longer use the stream, as
-  // per the filter contract.
-  processing_complete_ = true;
+void Filter::closeStream() {
   if (stream_) {
+    ENVOY_LOG(debug, "Calling close on stream");
     if (stream_->close()) {
       stats_.streams_closed_.inc();
     }
+    stream_.reset();
```
Comment (chaoqin-li1123) [inline source/extensions/filters/http/ext_proc/ext_proc.cc:74]: We are actually on the call stack of RawAsyncStream here, the reason why we can free the AsyncStream stream_ is that AsyncStream doesn't own the RawAsyncStream, and the RawAsyncStream will call cleanup() later to schedule a defer delete of itself. maybe document this hack.

**Comment by chaoqin-li1123** [comment]
The CI is passing, thanks for the quick fix!

**Comment by gbrail** [comment]
New PR message -- LMK if this works for the maintainers or if you'd all like a new PR:

Commit Message: Cache the gRPC client to reduce latency in the ext_proc filter.
Additional Description:
Risk Level: Low. Using a cache already used for other gRPC-based filters.
Testing: Existing test coverage is good.
Docs Changes:
Release Notes:

**Review by snowp — APPROVED** [review]
LGTM, thanks!

**Pull request merged**